The solution does not match Pillow in some cases. I just added Python test against Pillow library.
```
======================================================================
FAIL: test_resize_pillow (test_imgproc.Imgproc_Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/openvino/opencv/modules/imgproc/misc/python/test/test_imgproc.py", line 42, in test_resize_pillow
    self.assertTrue(status)
AssertionError: False is not true

----------------------------------------------------------------------
```

--- a/modules/imgproc/misc/python/test/test_imgproc.py
+++ b/modules/imgproc/misc/python/test/test_imgproc.py
@@ -45,4 +45,4 @@ class Imgproc_Tests(NewOpenCVTests):
             ocv_result = cv.resize(r, dsize=ts, interpolation=cv.INTER_NEAREST_EXACT)
             status = np.all(pil_result == ocv_result)
             print(ts, status)
-            self.assertTrue(status)
+            self.assertTrue(status, "resize result differs from Pillow for target size %s" % (ts,))

--- a/modules/imgproc/misc/python/test/test_imgproc.py
+++ b/modules/imgproc/misc/python/test/test_imgproc.py
@@ -4,7 +4,10 @@ from __future__ import print_function
 
 import numpy as np
 import cv2 as cv
-from PIL import Image
+try:
+    from PIL import Image
+except ImportError:
+    Image = None
 
 from tests_common import NewOpenCVTests
 
@@ -31,6 +34,9 @@ class Imgproc_Tests(NewOpenCVTests):
         self.assertLess(cv.norm(img_blur0 - img_blur1, cv.NORM_INF), eps)
 
     def test_resize_pillow(self):
+        if Image is None:
+            self.skipTest("Pillow is not available")
+
         r = np.random.randint(0, 255, size=(128, 147, 3), dtype="uint8")
         target_size=[(128,128), (129,129), (160,160)]
         for ts in target_size:

--- a/modules/imgproc/src/resize.cpp
+++ b/modules/imgproc/src/resize.cpp
@@ -1185,7 +1185,7 @@ public:
             uchar* D = dst.ptr(y);
             // Fixed-point coordinate mapping: floor((y + 0.5) * src_height / dst_height)
             // Using integer arithmetic: ((y * 2 + 1) * src_height) / (dst_height * 2)
-            int sy = std::min((int)(((int64_t)(y * 2 + 1) * src_height) / (dst_height * 2)), ssize.height-1);
+            int sy = std::min((int)((((int64_t)y * 2 + 1) * src_height) / ((int64_t)dst_height * 2)), ssize.height-1);
             const uchar* S = src.ptr(sy);
 
             int x = 0;
@@ -1277,7 +1277,7 @@ static void resizeNN_bitexact( const Mat& src, Mat& dst, double /*fx*/, double /
     // Using integer arithmetic to guarantee bit-exact results across platforms.
     for( int x = 0; x < dsize.width; x++ )
     {
-        x_ofse[x] = std::min((int)(((int64_t)(x * 2 + 1) * ssize.width) / (dsize.width * 2)), ssize.width-1);
+        x_ofse[x] = std::min((int)((((int64_t)x * 2 + 1) * ssize.width) / ((int64_t)dsize.width * 2)), ssize.width-1);
     }
     Range range(0, dsize.height);
     resizeNN_bitexactInvoker invoker(src, dst, x_ofse, ssize.height, dsize.height);

--- a/modules/imgproc/src/resize.cpp
+++ b/modules/imgproc/src/resize.cpp
@@ -1173,19 +1173,17 @@ resizeNN( const Mat& src, Mat& dst, double fx, double fy )
 class resizeNN_bitexactInvoker : public ParallelLoopBody
 {
 public:
-    resizeNN_bitexactInvoker(const Mat& _src, Mat& _dst, int* _x_ofse, int _src_height, int _dst_height)
-        : src(_src), dst(_dst), x_ofse(_x_ofse), src_height(_src_height), dst_height(_dst_height) {}
+    resizeNN_bitexactInvoker(const Mat& _src, Mat& _dst, int* _x_ofse, int* _y_ofse)
+        : src(_src), dst(_dst), x_ofse(_x_ofse), y_ofse(_y_ofse) {}
 
     virtual void operator() (const Range& range) const CV_OVERRIDE
     {
-        Size ssize = src.size(), dsize = dst.size();
+        Size dsize = dst.size();
         int pix_size = (int)src.elemSize();
         for( int y = range.start; y < range.end; y++ )
         {
             uchar* D = dst.ptr(y);
-            // Fixed-point coordinate mapping: floor((y + 0.5) * src_height / dst_height)
-            // Using integer arithmetic: ((y * 2 + 1) * src_height) / (dst_height * 2)
-            int sy = std::min((int)((((int64_t)y * 2 + 1) * src_height) / ((int64_t)dst_height * 2)), ssize.height-1);
+            int sy = y_ofse[y];
             const uchar* S = src.ptr(sy);
 
             int x = 0;
@@ -1260,9 +1258,22 @@ private:
     const Mat& src;
     Mat& dst;
     int* x_ofse;
-    const int src_height;
-    const int dst_height;
+    int* y_ofse;
 };
+
+static void resizeNN_bitexact_tab(int src_dim, int dst_dim, int* ofse)
+{
+    // Match Pillow's nearest-neighbor resize path: start at half a pixel,
+    // truncate the current coordinate, then increment by scale. softdouble
+    // keeps the IEEE-754 rounding deterministic across platforms.
+    const softdouble scale = softdouble(src_dim) / softdouble(dst_dim);
+    softdouble f = scale * softdouble(0.5);
+    for( int i = 0; i < dst_dim; i++ )
+    {
+        ofse[i] = std::min(cvTrunc(f), src_dim-1);
+        f += scale;
+    }
+}
 
 static void resizeNN_bitexact( const Mat& src, Mat& dst, double /*fx*/, double /*fy*/ )
 {
@@ -1270,17 +1281,16 @@ static void resizeNN_bitexact( const Mat& src, Mat& dst, double /*fx*/, double /
 
     cv::utils::BufferArea area;
     int* x_ofse = 0;
+    int* y_ofse = 0;
     area.allocate(x_ofse, dsize.width, CV_SIMD_WIDTH);
+    area.allocate(y_ofse, dsize.height, CV_SIMD_WIDTH);
     area.commit();
 
-    // Fixed-point coordinate mapping: floor((x + 0.5) * src_width / dst_width)
-    // Using integer arithmetic to guarantee bit-exact results across platforms.
-    for( int x = 0; x < dsize.width; x++ )
-    {
-        x_ofse[x] = std::min((int)((((int64_t)x * 2 + 1) * ssize.width) / ((int64_t)dsize.width * 2)), ssize.width-1);
-    }
+    resizeNN_bitexact_tab(ssize.width, dsize.width, x_ofse);
+    resizeNN_bitexact_tab(ssize.height, dsize.height, y_ofse);
+
     Range range(0, dsize.height);
-    resizeNN_bitexactInvoker invoker(src, dst, x_ofse, ssize.height, dsize.height);
+    resizeNN_bitexactInvoker invoker(src, dst, x_ofse, y_ofse);
     parallel_for_(range, invoker, dst.total()/(double)(1<<16));
 }
 

--- a/modules/imgproc/test/test_resize_bitexact.cpp
+++ b/modules/imgproc/test/test_resize_bitexact.cpp
@@ -253,7 +253,7 @@ TEST(Resize_Bitexact, NearestExact_PillowCompat)
         mapping.resize(dst_dim);
         for (int i = 0; i < dst_dim; i++)
         {
-            mapping[i] = std::min((int)(((int64_t)(i * 2 + 1) * src_dim) / (dst_dim * 2)), src_dim - 1);
+            mapping[i] = std::min((int)((((int64_t)i * 2 + 1) * src_dim) / ((int64_t)dst_dim * 2)), src_dim - 1);
         }
     };
 

--- a/modules/imgproc/test/test_resize_bitexact.cpp
+++ b/modules/imgproc/test/test_resize_bitexact.cpp
@@ -240,20 +240,18 @@ TEST(Resize_Bitexact, Nearest8U)
     }
 }
 
-// Regression test for #28429: INTER_NEAREST_EXACT uses fixed-point integer
-// arithmetic for center-of-pixel coordinate mapping. The formula
-// floor((i + 0.5) * src / dst) is computed as ((i*2+1)*src) / (dst*2)
-// using int64_t to avoid overflow and guarantee bit-exact results
-// across all platforms without hardware FP dependency.
+// Regression test for #28429: INTER_NEAREST_EXACT matches Pillow's
+// center-of-pixel nearest-neighbor mapping.
 TEST(Resize_Bitexact, NearestExact_PillowCompat)
 {
-    // Fixed-point center-of-pixel mapping: floor((i + 0.5) * src_dim / dst_dim)
-    // Integer form: ((i * 2 + 1) * src_dim) / (dst_dim * 2)
     auto center_pixel_map = [](int src_dim, int dst_dim, std::vector<int>& mapping) {
+        softdouble scale = softdouble(src_dim) / softdouble(dst_dim);
+        softdouble f = scale * softdouble(0.5);
         mapping.resize(dst_dim);
         for (int i = 0; i < dst_dim; i++)
         {
-            mapping[i] = std::min((int)((((int64_t)i * 2 + 1) * src_dim) / ((int64_t)dst_dim * 2)), src_dim - 1);
+            mapping[i] = std::min(cvTrunc(f), src_dim - 1);
+            f += scale;
         }
     };
 

